### ISSUE TITLE
IPv6 Compliance RFC4861 + RFC4862: Router Advertisement Processing, On-link determination (Host Only) [v6LC.2.2.19]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#28438]
### systemd version the issue has been seen with

250

### Used distribution

Embedded Linux - Debian-based

### Linux kernel version used

5.15.71-rt51+gc36e774d0d9a

### CPU architectures issue was seen on

aarch64

### Expected behaviour you didn't see

**Purpose**: Verify that a host properly rejects an invalid prefix length, however the prefix length is still valid for on-link determination when the on-link flag is true.

Despite the RA's prefix length being invalid, the L-flag is set. Therefore the host must respond to echo-requests with a reply on-link.

From [RFC4861](https://www.rfc-editor.org/rfc/rfc4861.txt)
```
6.3.4.  Processing Received Router Advertisements
However, the prefix length is still valid for on-link determination when
combined with other flags in the prefix option.
```

From [RFC4862](https://www.rfc-editor.org/rfc/rfc4862.txt), section 5.5.3

```
      It is the responsibility of the system administrator to ensure
      that the lengths of prefixes contained in Router Advertisements
      are consistent with the length of interface identifiers for that
      link type.  It should be noted, however, that this does not mean
      the advertised prefix length is meaningless.  In fact, the
      advertised length has non-trivial meaning for on-link
      determination in [RFC4861] where the sum of the prefix length and
      the interface identifier length may not be equal to 128.  Thus, it
      should be safe to validate the advertised prefix length here, in
      order to detect and avoid a configuration error specifying an
      invalid prefix length in the context of address autoconfiguration.

```

### Unexpected behaviour you saw

Did not receive an echo reply on-link.

### Steps to reproduce the problem

Please find attached a zip file containing two PCAP files, `legacy` and `networkd`.  The `legacy` PCAP shows the correct behavior.  It was recorded when our host was configured using Debian's legacy `/etc/network/interfaces` file and `networkd` was disabled and stopped.  The `networkd` PCAP shows incorrect behavior. It was recorded after configuring `networkd` to manage the interface.

The incorrect behaviour is shown under the `networkd` folder the PCAP file `v6LC_2_2_19-Net0-20230707043454-filtered.pcapng` shows packet 8 is a ping request, for which there is no reply.

Compare that with the correct behaviour shown under the `legacy` folder. The PCAP file `v6LC_2_2_19-Net0-20230705084536.pcapng` shows packet 7 is a ping request to which the host replies.

[2_2_19 - Router Advertisement Processing, On-link determination.zip](https://github.com/systemd/systemd/files/12075649/2_2_19.-.Router.Advertisement.Processing.On-link.determination.zip)

**Comment by yuwata** [comment]
For this issue, what configuration is missing? or what is configured wrongly? Please provide difference of e.g. `ip route` compared with the 'legacy' case.

**Comment by LiveFreeAndRoam** [comment]
@yuwata, I'll get you some additional information.  If I can create a PCAP that reproduces the issue, would you be able to step through it using `tcpreplay`?  If you've never used `tcpreplay` before, I can give you some more instructions.

**Comment by LiveFreeAndRoam** [comment]
Wait on, I just reran 2.2.19 using the systemd ToT and it is passing.  I then repeated the test on systemd v253.5 and it still passes. 
 I'm not sure what fixed it between 250 and 253.5, but I can call this one fixed.

I'll close this issue.
